File: examples/mobilenet_ssd.cpp
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * License); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * AS IS BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*
 * Copyright (c) 2020, OPEN AI LAB
 * Author: sqfu@openailab.com
 */

#include <unistd.h>
#include <iostream>
#include <iomanip>
#include <string>
#include <vector>
#include <sys/time.h>
#include "tengine_operations.h"
#include "tengine_cpp_api.h"
#include "common.hpp"

static std::string gExcName{""};

struct Box
{
    float x0;
    float y0;
    float x1;
    float y1;
    int class_idx;
    float score;
};

void get_input_data_ssd(std::string& image_file, float* input_data, int img_h, int img_w)
{
    float mean[3] = {127.5, 127.5, 127.5};
    float scales[3] = {0.007843, 0.007843, 0.007843};
    image img = imread(image_file.c_str(), img_w, img_h, mean, scales, CAFFE);    
    memcpy(input_data, img.data, sizeof(float)*3*img_w*img_h);  
    free_image(img);    
}

void post_process_ssd(std::string& image_file, float threshold, float* outdata, int num)
{
    const char* class_names[] = {"background", "aeroplane", "bicycle",   "bird",   "boat",        "bottle",
                                 "bus",        "car",       "cat",       "chair",  "cow",         "diningtable",
                                 "dog",        "horse",     "motorbike", "person", "pottedplant", "sheep",
                                 "sofa",       "train",     "tvmonitor"};

    image im = imread(image_file.c_str());

    int raw_h = im.h;
    int raw_w = im.w;
    std::vector<Box> boxes;
    printf("detect result num: %d \n", num);
    for(int i = 0; i < num; i++)
    {
        if(outdata[1] >= threshold)
        {
            Box box;
            box.class_idx = outdata[0];
            box.score = outdata[1];
            box.x0 = outdata[2] * raw_w;
            box.y0 = outdata[3] * raw_h;
            box.x1 = outdata[4] * raw_w;
            box.y1 = outdata[5] * raw_h;
            boxes.push_back(box);
            printf("%s\t:%.0f%%\n", class_names[box.class_idx], box.score * 100);
            printf("BOX:( %d , %d ),( %d , %d )\n", (int)box.x0, (int)box.y0, (int)box.x1, (int)box.y1);
        }
        outdata += 6;
    }
    for(int i = 0; i < ( int )boxes.size(); i++)
    {
        Box box = boxes[i];

        std::ostringstream score_str;
        score_str << box.score * 100;
        std::string labelstr = std::string(class_names[box.class_idx]) + " : " + score_str.str();

        put_label(im, labelstr.c_str(), 0.02, box.x0, box.y0, 255, 255, 125);
        draw_box(im, box.x0, box.y0, box.x1, box.y1, 2, 125, 0, 125);
    }

    save_image(im, "tengine_example_out");
    free_image(im);
    std::cout << "======================================\n";
    std::cout << "[DETECTED IMAGE SAVED]:\t"
              << "Mobilenet_SSD"
              << "\n";
    std::cout << "======================================\n";
}

void show_usage()
{
    std::cout << "[Usage]: " << gExcName << " [-h]\n"
              << "   [-m model_file] [-i image_file]\n";
}

int main(int argc, char* argv[])
{
    gExcName = std::string(argv[0]);
    int ret = -1;
    std::string model_file;
    std::string image_file;
    const char* device = nullptr;

    int res;
    while((res = getopt(argc, argv, "m:i:hd:")) != -1)
    {
        switch(res)
        {
            case 'm':
                model_file = optarg;
                break;
            case 'i':
                image_file = optarg;
                break;
            case 'd':
                device = optarg;
                break;
            case 'h':
                show_usage();
                return 0;
            default:
                break;
        }
    }

    if( model_file.empty() )
    {
        std::cerr << "Error: model file not specified!" << std::endl;
        show_usage();
        return -1;
    }

    if(image_file.empty())
    {
        std::cerr << "Error: image file not specified!" << std::endl;
        show_usage();
        return -1;
    }

    tengine::Net somenet;
    tengine::Tensor input_tensor;
    tengine::Tensor output_tensor;
	
    if(request_tengine_version("0.9") != 1)
    {
        std::cout << " request tengine version failed\n";
        return 1;
    }
    // check file
    if((!check_file_exist(model_file) or !check_file_exist(image_file)))
    {
        return 1;
    }

    /* load model */
    somenet.load_model(NULL, "tengine", model_file.c_str());

    // input
    int img_h = 300;
    int img_w = 300;
    int img_size = img_h * img_w * 3;

    float mean[3] = {127.5, 127.5, 127.5};

    float scales = 0.007843;

    int repeat_count = 1;
    const char* repeat = std::getenv("REPEAT_COUNT");

    if(repeat)
        repeat_count = std::strtoul(repeat, NULL, 10);

    /* prepare input data */
    input_tensor.create(img_w, img_h, 3);

    get_input_data(image_file.c_str(), (float* )input_tensor.data, img_h, img_w, mean, scales);

    /* forward */
    somenet.input_tensor(0, 0, input_tensor);

    struct timeval t0, t1;
    float total_time = 0.f;
    float min_time = __DBL_MAX__;
    float max_time = -__DBL_MAX__;
    for(int i = 0; i < repeat_count; i++)
    {
        gettimeofday(&t0, NULL);
        somenet.run();
        gettimeofday(&t1, NULL);
        float mytime = ( float )((t1.tv_sec * 1000000 + t1.tv_usec) - (t0.tv_sec * 1000000 + t0.tv_usec)) / 1000;
        total_time += mytime;
        min_time = std::min(min_time, mytime);
        max_time = std::max(max_time, mytime);
    }
    std::cout << "--------------------------------------\n";
    std::cout << "\nRepeat " << repeat_count << " times, avg time per run is " << total_time / repeat_count << " ms\n" << "max time is " << max_time << " ms, min time is " << min_time << " ms\n";

    //tensor_t out_tensor = get_graph_output_tensor(graph, 0, 0);    //"detection_out");
    /* get result */
    somenet.extract_tensor(0, 0, output_tensor);

    float* outdata = ( float* )(output_tensor.data);

    float show_threshold = 0.5;

    post_process_ssd(image_file, show_threshold, outdata, output_tensor.h);

    return 0;
}

File: examples/mobilenet_ssd.py
from tengine import tg
import numpy as np
import cv2
import argparse
import os
import time


DEFAULT_MAX_BOX_COUNT = 100
DEFAULT_REPEAT_COUNT = 1
DEFAULT_THREAD_COUNT = 1
DEFAULT_IMG_H = 300
DEFAULT_IMG_W = 300
DEFAULT_SCALE = 0.008
DEFAULT_MEAN1 = 127.5
DEFAULT_MEAN2 = 127.5
DEFAULT_MEAN3 = 127.5
SHOW_THRESHOLD = 0.5

parser = argparse.ArgumentParser(description='mobilenet_ssd')
parser.add_argument('-m', '--model', default='./models/mobilenet_ssd.tmfile', type=str)
parser.add_argument('-i', '--image', default='./images/dog.jpg', type=str)
parser.add_argument('-r', '--repeat_count', default=f'{DEFAULT_REPEAT_COUNT}', type=str)
parser.add_argument('-t', '--thread_count', default=f'{DEFAULT_THREAD_COUNT}', type=str)


def get_current_time():
    return time.time() * 1000

def draw_box(img, x1, y1, x2, y2, w=2, r=125, g=0, b=125):
    im_h,im_w,im_c = img.shape
    #print("draw_box", im_h, im_w, x1, x2, y1, y2)
    x1 = np.clip(x1, 0, im_w)
    x2 = np.clip(x2, 0, im_w)
    y1 = np.clip(y1, 0, im_h)
    y2 = np.clip(y2, 0, im_h)
    img[y1:y2, x1:x1+w] = [r, g, b]
    img[y1:y2, x2:x2+w] = [r, g, b]
    img[y1:y1+w, x1:x2] = [r, g, b]
    img[y2:y2+w, x1:x2] = [r, g, b]


def main(args):
    class_name=["background", "aeroplane", "bicycle",   "bird",   "boat",        "bottle",
                "bus",        "car",       "cat",       "chair",  "cow",         "diningtable",
                "dog",        "horse",     "motorbike", "person", "pottedplant", "sheep",
                "sofa",       "train",     "tvmonitor"]
    image_file = args.image
    tm_file = args.model
    assert os.path.exists(tm_file), 'Model: {tm_file} not found'
    assert os.path.exists(image_file), 'Image: {image_file} not found'

    img_h,img_w =map(int,(DEFAULT_IMG_H, DEFAULT_IMG_W))
    scale = list(map(float,(DEFAULT_SCALE,DEFAULT_SCALE,DEFAULT_SCALE)))
    mean_value = list(map(float,(DEFAULT_MEAN1,DEFAULT_MEAN2,DEFAULT_MEAN3)))
    show_threshold = float(SHOW_THRESHOLD)
    assert len(mean_value) == 3, 'The number of mean_value should be 3, e.g. 104.007,116.669,122.679'
    img_mean = np.array(mean_value).reshape((1, 1, 3))
    data = cv2.imread(image_file)
    h=data.shape[0]
    w=data.shape[1]
    data = cv2.resize(data, (img_w, img_h))
    data = ((data - img_mean) * DEFAULT_SCALE).astype(np.float32)
    data = data.transpose((2, 0, 1))
    assert data.dtype == np.float32
    data = data.copy()

    graph = tg.Graph(None, 'tengine', tm_file)
    input_tensor = graph.getInputTensor(0, 0)
    dims = [1, 3, img_h, img_w]
    input_tensor.shape = dims
    graph.preRun()
    input_tensor.buf = data
    graph.run(1) # 1 is blocking
    output_tensor = graph.getOutputTensor(0, 0)
    output = output_tensor.getNumpyData()
    num=output.shape[1]
    output=np.squeeze(output)

    print('detect result num: %d'%num)
    BOX=[]
    for i in range(int(num)):
        box={}
        score = output[i][1]
        if score>show_threshold:
            class_idx = output[i][0]
            x0 = output[i][2] * w
            y0 = output[i][3] * h
            x1 = output[i][4] * w
            y1 = output[i][5] * h
            box["x0"]=int(x0)
            box["y0"]=int(y0)
            box["x1"]=int(x1)
            box["y1"]=int(y1)
            print('class_names[%d]:%s\tscore: %.2f%%'%(int(class_idx), class_name[int(class_idx)],(score*100)),'  BOX:( %d , %d ),( %d , %d )'%(x0,y0,x1,y1))
        BOX.append(box)
    
    data_out = cv2.imread(image_file)
    for i in range(int(num)):
        draw_box(img=data_out, x1= BOX[i]["x0"], y1=BOX[i]["y0"], x2=BOX[i]["x1"], y2=BOX[i]["y1"], w=2, r=125, g=0, b=125)
        
    out_path="./mobilenet_ssd_out.jpg"
    cv2.imwrite(out_path,data_out)
    print("save_img path:",out_path)

if __name__ == '__main__':
    args = parser.parse_args()
    main(args)


